timemachine/fe/protocol_refinement.py

```diff
@@ -0,0 +1,30 @@
+from typing import Callable, List, Optional, TypeVar
+
+_T = TypeVar("_T")
+
+
+def greedy_bisection_step(
+    protocol: List[_T],
+    local_cost: Callable[[_T, _T], float],
+    make_intermediate: Callable[[_T, _T], _T],
+    callback: Optional[Callable[[float, _T, _T, _T], None]] = None,
+) -> List[_T]:
+    assert len(protocol) >= 2
+
+    adjacent_pairs = zip(protocol, protocol[1:])
+    max_cost, s1_idx, (s1, s2) = max((local_cost(s1, s2), idx, (s1, s2)) for idx, (s1, s2) in enumerate(adjacent_pairs))
+    new_state = make_intermediate(s1, s2)
+
+    if callback:
+        callback(max_cost, s1, s2, new_state)
```

Comment:
nit[flexibility]: Would it be useful for this to retain the whole list of local costs, not just the max cost?

Reply:
Addressed in https://github.com/proteneer/timemachine/pull/959/commits/0c05a13ee28e0fd0adf48e6e0523e14d17e2b2a5